columnflow/production/cms/dy.py

```diff
@@ -0,0 +1,459 @@
+        hpara_var = self.recoil_unc_corrector.evaluate(
+            self.dy_recoil_config.era,
+            self.dy_recoil_config.order,
+            njet,
+            events.gen_dilepton_all.pt,
+            "Hpara",
+            H_para,
+            syst,
+        )
+        hperp_var = self.recoil_unc_corrector.evaluate(
+            self.dy_recoil_config.era,
+            self.dy_recoil_config.order,
+            njet,
+            events.gen_dilepton_all.pt,
+            "Hperp",
+            H_perp,
+            syst,
+        )
+        # Reconstruct the corrected H vector in the full boson frame
+        H_corr_x = hpara_var * np.cos(full.phi) - hperp_var * np.sin(full.phi)
+        H_corr_y = hpara_var * np.sin(full.phi) + hperp_var * np.cos(full.phi)
+        # Reconstruct the MET variation: MET_var = -H_corr - vis
+        met_var_x = -H_corr_x - vis.x
+        met_var_y = -H_corr_y - vis.y
+        met_var_pt = np.sqrt(met_var_x**2 + met_var_y**2)
+        met_var_phi = np.arctan2(met_var_y, met_var_x)
+        events = set_ak_column(
+            events, f"RecoilCorrMET.pt_{syst}", met_var_pt, value_type=np.float32,
+        )
+        events = set_ak_column(
+            events, f"RecoilCorrMET.phi_{syst}", met_var_phi, value_type=np.float32,
+        )
+
+    return events
+
+
+@recoil_corrections.init
+def recoil_corrections_init(self: Producer) -> None:
+    # the number of weights in partial run 3 is always 10
+    if self.config_inst.campaign.x.year not in (2022, 2023):
+        raise NotImplementedError(
+            f"campaign year {self.config_inst.campaign.x.year} is not yet supported by {self.cls_name}",
+        )
+
+    self.systematics = ["RespUp", "RespDown", "ResolUp", "ResolDown"]
+    # register dynamically systematics
+    for syst in self.systematics:
+        self.produces.add(f"RecoilCorrMET.pt_{syst}")
+        self.produces.add(f"RecoilCorrMET.phi_{syst}")
+
+
+@recoil_corrections.requires
+def recoil_corrections_requires(self: Producer, reqs: dict) -> None:
+    # Ensure that external files are bundled.
+    if "external_files" in reqs:
+        return
+    from columnflow.tasks.external import BundleExternalFiles
+
+    reqs["external_files"] = BundleExternalFiles.req(self.task)
+
+
+@recoil_corrections.setup
+def recoil_corrections_setup(
+    self: Producer, reqs: dict, inputs: dict, reader_targets: InsertableDict,
+) -> None:
+    """
+    Setup the recoil corrections by loading the CorrectionSet via correctionlib.
+    The external recoil correction file should be provided as external_files.recoil.
+    """
+    bundle = reqs["external_files"]
+
+    import correctionlib
+
+    correctionlib.highlevel.Correction.__call__ = (
+        correctionlib.highlevel.Correction.evaluate
+    )
+
+    # Load the correction set from the external file.
+    correction_set = correctionlib.CorrectionSet.from_string(
+        self.get_dy_recoil_file(bundle.files).load(formatter="gzip").decode("utf-8"),
+    )
```

Comment:
Using the new `load_correction_set` helper.

```suggestion
    correction_set = load_correction_set(self.get_dy_recoil_file(bundle.files))
```